add new cookbook section with usecases (#5757)
dbt-labs/docs.getdbt.com: website/docs/guides/sl-cookbook-recipes.md

---
title: dbt Semantic Layer cookbook recipes

> id: sl-cookbook-recipes

Author: ```suggestion
id: sl-cookbook
```

description: "The dbt dbt Semantic Layer cookbook recipes are a collection of scenario-based, real-world examples for building with the dbt Semantic Layer."
displayText: Cookbook recipes offer practical, scenario-based examples for semantic models and metrics with the dbt Semantic Layer.
hoverSnippet: Read a collection of scenario-based, real-world examples for building metrics and semantic models with the dbt Semantic Layer.
# time_to_complete: '30 minutes' commenting out until we test
hide_table_of_contents: true
tags: ['Semantic Layer','dbt Cloud']
level: 'Intermediate'
recently_updated: true
---

<div style={{maxWidth: '900px'}}>

> ## How to calculate active users and active users WoW growth

Author: need to add an intro h2 header before diving into this

Use this cookbook recipe to understand how to calculate “Active users” and “Active users (Week-over-Week growth (WoW) growth), as well as other time-related attributes in a database with dbt using YAML configurations.

- [Objectives](#objectives)

> - [Use cases](#use-case)

Reviewer: Should this be singular as the title is singular? or should the title be plural? 🤔 
```suggestion
- [Use case](#use-case)
```

- Learn how to define and calculate metrics for active users and their WoW growth. As well as utility models for fiscal periods and other date-related flags.

### Use case

Track and analyze user activity on a daily basis, including page views, time-on-site, and other key metrics. This helps in monitoring user engagement and understanding trends over time.

:::tip
💡 Note that we’re using the [parallel subfolder approach](/best-practices/how-we-build-our-metrics/semantic-layer-7-semantic-structure#files-and-folders) for better organization in large projects, which means you’re creating a semantic model folder containing all your semantic layer code. 

Alternatively, you can use the [one-YAML-per-marts-model approach](/best-practices/how-we-build-our-metrics/semantic-layer-7-semantic-structure#files-and-folders) which puts documentation, data tests, unit tests, semantic models, and metrics into a unified file corresponding to a dbt-modeled mart.
:::

### Step 1: **Define semantic models**

1. Create a YAML file under the `semantic_models` folder: (`models/semantic_models/sem_user_activity_daily.yml`) to create a semantic model that tracks daily user activity. 

>   This semantic model ref’s the `fct_user_activity_daily` model.

Reviewer: ```suggestion
  This semantic model refs the `fct_user_activity_daily` model.
```

    model: ref('all_days')
    defaults:
      agg_time_dimension: date_day

    entities:
      - name: date
        type: primary
        expr: date_day
      - name: date_month
        type: foreign
        expr: month_start_date

    dimensions:
      - name: date_day
        type: time
        type_params:
          time_granularity: day
      - name: date_month
        type: time
        type_params:
          time_granularity: month
      - name: fiscal_quarter
        type: categorical
      - name: fiscal_year
        type: categorical
      - name: is_end_of_week
        type: categorical
      - name: is_yesterday
        type: categorical
```
</File>

### Step 2: Create metrics

1. Create a YAML file under the `metrics` folder: (`models/metrics/sem_user_activity_daily.yml`) to create a metric that tracks active users. 
   
  Note that if you’re using the one-YAML-file-per-mart, then save this as `active_users.yml` under the models folder.

<File name="models/metrics/sem_user_activity_daily.yml">

```yaml
metrics:
  - name: users_active
    label: 'Users: active'

>     description: >

Reviewer: I've been encouraging the Data team to use `|` over `>` as the `|` takes newlines into consideration within the markdown snippet.
```suggestion
    description: |
```

<File name="models/semantic_models/sem_revenue_models.yml">

```yaml
semantic_models:
  - name: sem_revenue
    description: >
      This model aggregates revenue data, providing metrics for business to analyze the financial 
      health of the fictional Santiago's snacks org.
    model: ref('fct_revenue')
    defaults:
      agg_time_dimension: date_month
    entities:
      - name: date
        type: primary
        expr: date_month
    measures:
      - name: mrr
        agg: sum
        expr: monthly_recurring_revenue
        agg_time_dimension: date_month
    dimensions:
      - name: date_day
        type: time
        type_params:
          time_granularity: day
      - name: date_month
        type: time
        type_params:
          time_granularity: month
      - name: is_last_day_of_month
        type: categorical

```
</File>

### Step 2: Define the ARR metric
Create another YAML file under the metrics folder: (`models/metrics/sem_revenue_metrics.yml`) to define the ARR metric.

<File name="models/metrics/sem_revenue_metrics.yml">

```yaml
metrics:
  - name: arr
    label: 'ARR'
    description: >

>       annual recurring revenue calculated based on MRR.

Reviewer: ```suggestion
      Annual recurring revenue calculated based on MRR.
```

- [Step 2: Create metrics](#step-2-create-metrics)
- [Step 3: Test your metrics](#step-3-test-your-metrics)

### Objectives

- Understand how to configure and use semantic models in dbt to track and report daily user activity.
- Learn how to define and calculate metrics for active users and their WoW growth. As well as utility models for fiscal periods and other date-related flags.

### Use case

Track and analyze user activity on a daily basis, including page views, time-on-site, and other key metrics. This helps in monitoring user engagement and understanding trends over time.

:::tip
💡 Note that we’re using the [parallel subfolder approach](/best-practices/how-we-build-our-metrics/semantic-layer-7-semantic-structure#files-and-folders) for better organization in large projects, which means you’re creating a semantic model folder containing all your semantic layer code. 

Alternatively, you can use the [one-YAML-per-marts-model approach](/best-practices/how-we-build-our-metrics/semantic-layer-7-semantic-structure#files-and-folders) which puts documentation, data tests, unit tests, semantic models, and metrics into a unified file corresponding to a dbt-modeled mart.
:::

> ### Step 1: **Define semantic models**

Author: ```suggestion
### Step 1: Define semantic models
```

1. Create a YAML file under the `semantic_models` folder: (`models/semantic_models/sem_user_activity_daily.yml`) to create a [semantic model](/docs/build/semantic-models) that tracks daily user activity. 
   This semantic model refs the `fct_user_activity_daily` model.

   Note that if you’re using the one-YAML-file-per-mart, then save this as `user_activity_daily.yml` under the models folder.

    <File name="models/semantic_models/sem_user_activity_daily.yml">

    ```yaml
    semantic_models:
      - name: sem_user_activity_daily
        description: |
          This model reports on daily user activity. Activity is broken down into
          pageview + time-on-site metrics for each use case that is addressed in
          dbt Cloud (ie. Deployment and Development)
        model: ref('fct_user_activity_daily')
        defaults:
          agg_time_dimension: date_day
    ```
    </File>

> 2. Add the following YAML configurations in that same semantic model file:

Author: will make sure all yaml examples align with paragraphs/text

```yaml
semantic_models:
  - name: sem_customer_acquisition
    description: |
      This model aggregates customer acquisition data, providing metrics for business to analyze the cost of acquiring new customers.
    model: ref('fct_customer_acquisition')
    defaults:
      agg_time_dimension: date_month
    entities:
      - name: date
        type: primary
        expr: date_month
    measures:
      - name: total_acquisition_cost
        agg: sum
        expr: acquisition_cost
        agg_time_dimension: date_month
      - name: new_customers
        agg: sum
        expr: customer_count
        agg_time_dimension: date_month
    dimensions:
      - name: date_month
        type: time
        type_params:
          time_granularity: month
      - name: acquisition_channel
        type: categorical
```

</File>

### Step 2: Define the CAC metric
Create another YAML file under the metrics folder: (`models/metrics/sem_customer_acquisition_metrics.yml`) to define the CAC metric.

<File name="models/metrics/sem_customer_acquisition_metrics.yml">

```yaml
metrics:
  - name: cac
    description: "Customer acquisition cost"
    label: 'CAC'        

>     type: ratio

Reviewer 2: `ratio` metrics require metrics as inputs for numerator, denominator. So we will need to define those metrics too (right now it seems like we are trying to use the measures)

    type_params:
      numerator: total_acquisition_cost
      denominator: new_customers
```

</File>

### Step 3: Validate and query metric

1. Confirm and validate the new metric definition in your development by running the following command:
   
   ```bash
   dbt parse
   ```  

2. Query your metrics to confirm the results are as expected. Run the following command:
   
   ```bash
   dbt sl query --metrics cac --group-by metric_time__year --where "metric_time__year >= '2023'" --order-by metric_time__year
   ```
<!-- Cookbook 4 -->

> ## How to join multiple fact metrics

Author: hey @patkearns10  - would love your feedback and perspective on this join metrics section if you have time!

Reviewer 2: Couple quick changes, but otherwise looks good to me!

Alternatively, you can use the [one-YAML-per-marts-model approach](/best-practices/how-we-build-our-metrics/semantic-layer-7-semantic-structure#files-and-folders) which puts documentation, data tests, unit tests, semantic models, and metrics into a unified file corresponding to a dbt-modeled mart.
:::

### Step 1: Define semantic models

1. Create a YAML file under the `semantic_models` folder (`models/semantic_models/sem_sales.yml`) to define a semantic model for tracking sales. This model should reference the `fct_sales` model and include relevant entities, dimensions, and measures. Note that if you’re using the one-YAML-file-per-mart, then save this as `sales.yml` under the models folder.

<File name="models/semantic_models/sem_sales.yml">

```yaml
semantic_models:
  - name: sem_sales
    description: >
      This model reports on sales activities, focusing on store-level and date-level metrics.
    model: ref('fct_sales')
    defaults:
      agg_time_dimension: date_day
    
    entities:

>       - name: store_id

Reviewer 2: ```suggestion
      - name: store
```
